### ISSUE TITLE
High lock contention in Socket.select in JRuby 9.3

### ISSUE THREAD

**Issue by jsvd** [jruby/jruby#7805]
**Environment Information**
  
Provide at least:
* JRuby version (`jruby -v`) and command line (flags, JRUBY_OPTS, etc)
```
jruby 9.3.10.0 (2.6.8) 2023-02-01 107b2e6697 OpenJDK 64-Bit Server VM 17.0.6+10 on 17.0.6+10 +indy +jit [x86_64-linux]
```
```
-Xms8g -Xmx8g -Djava.awt.headless=true -Dfile.encoding=UTF-8 -Djruby.compile.invokedynamic=true -Djruby.jit.threshold=0 -Djruby.regexp.interruptible=true -XX:+HeapDumpOnOutOfMemoryError -Djava.security.egd=file:/dev/urandom -Dlog4j2.isThreadContextMapInheritable=true -Djdk.io.File.enableADS=true --add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED --add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED --add-exports=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED --add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED --add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED --add-opens=java.base/java.security=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.nio.channels=ALL-UNNAMED --add-opens=java.base/sun.nio.ch=ALL-UNNAMED --add-opens=java.management/sun.management=ALL-UNNAMED
```
* Operating system and platform: observed in SLES/Ubuntu/Debian kernels 4.x and 5.x

Other relevant info you may wish to add:

* Code runs inside Logstash: https://github.com/logstash-plugins/logstash-input-syslog/blob/main/lib/logstash/inputs/syslog.rb#L235-L249

#### Problem description

Multiple endpoints pushing syslog data to a central TCP server.
Since the upgrade from JRuby 9.2 to 9.3.10.0, multiple users of Logstash are observing very high cpu usage.
I have not yet been able to reproduce this locally in macos or a debian vm on GCP.
What users observe is that multiple threads are in a blocked state.

If everything is going well, what we see is a runnable thread:

```
"input|syslog|tcp" #134 daemon prio=5 os_prio=0 cpu=88879,49ms elapsed=3904,82s tid=0x00007f42e4098990 nid=0x73cd2 runnable  [0x00007f42c62e0000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.EPoll.wait(java.base@17.0.6/Native Method)
	at sun.nio.ch.EPollSelectorImpl.doSelect(java.base@17.0.6/EPollSelectorImpl.java:118)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(java.base@17.0.6/SelectorImpl.java:129)
	- locked <0x000000060efe8310> (a sun.nio.ch.Util$2)
	- locked <0x000000060efe82c0> (a sun.nio.ch.EPollSelectorImpl)
	at sun.nio.ch.SelectorImpl.select(java.base@17.0.6/SelectorImpl.java:146)
	at org.jruby.RubyThread.select(RubyThread.java:2132)
	- locked <0x000000060ecb7f68> (a java.lang.Object)
	at org.jruby.RubyThread.select(RubyThread.java:2053)
	at org.jruby.util.io.OpenFile.readInternal(OpenFile.java:1394)
	at org.jruby.util.io.OpenFile.fillbuf(OpenFile.java:1295)
	at org.jruby.util.io.OpenFile.getlineFast(OpenFile.java:1635)
	at org.jruby.RubyIO.getlineImpl(RubyIO.java:716)
	at org.jruby.RubyIO.access$100(RubyIO.java:122)
	at org.jruby.RubyIO$4.getline(RubyIO.java:2492)
	at org.jruby.RubyIO$4.getline(RubyIO.java:2487)
	at org.jruby.util.io.Getline.getlineCall(Getline.java:129)
	at org.jruby.util.io.Getline.getlineCall(Getline.java:39)
	at org.jruby.RubyIO.each(RubyIO.java:3533)
	at java.lang.invoke.LambdaForm$DMH/0x0000000801299000.invokeVirtual(java.base@17.0.6/LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/0x00000008012a0c00.invoke(java.base@17.0.6/LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/0x00000008012d2400.tryFinally(java.base@17.0.6/LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/0x000000080127d800.reinvoke(java.base@17.0.6/LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/0x000000080127dc00.guard(java.base@17.0.6/LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/0x000000080127d800.reinvoke(java.base@17.0.6/LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/0x000000080127dc00.guard(java.base@17.0.6/LambdaForm$MH)
	at java.lang.invoke.Invokers$Holder.linkToCallSite(java.base@17.0.6/Invokers$Holder)
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_6_dot_0.gems.logstash_minus_input_minus_syslog_minus_3_dot_6_dot_0.lib.logstash.inputs.syslog.RUBY$method$tcp_receiver$0(/usr/share/logstash/vendor/bundle/jruby/2.6.0/gems/logstash-input-syslog-3.6.0/lib/logstash/inputs/syslog.rb:235)
```

However, on a very high cpu usage scenario we see dozens of blocked threads, one per endpoint connecting to the TCP Server:

```
"input|syslog|tcp" #29570 daemon prio=5 os_prio=0 cpu=2022109.85ms elapsed=26655.76s tid=0x00007ff4b41b0a30 nid=0x8ed2 waiting for monitor entry  [0x00007ff4332b1000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.jruby.util.io.SelectorPool.returnToPool(SelectorPool.java:137)
	- waiting to lock <0x00007ff55de4da48> (a org.jruby.util.io.SelectorPool)
	at org.jruby.util.io.SelectorPool.put(SelectorPool.java:100)
	at org.jruby.RubyThread.select(RubyThread.java:2163)
	- locked <0x00007ff80da3cf38> (a java.lang.Object)
	at org.jruby.RubyThread.select(RubyThread.java:2053)
	at org.jruby.util.io.OpenFile.readInternal(OpenFile.java:1394)
	at org.jruby.util.io.OpenFile.fillbuf(OpenFile.java:1295)
	at org.jruby.util.io.OpenFile.getlineFast(OpenFile.java:1635)
	at org.jruby.RubyIO.getlineImpl(RubyIO.java:716)
	at org.jruby.RubyIO.access$100(RubyIO.java:122)
	at org.jruby.RubyIO$4.getline(RubyIO.java:2492)
	at org.jruby.RubyIO$4.getline(RubyIO.java:2487)
	at org.jruby.util.io.Getline.getlineCall(Getline.java:129)
	at org.jruby.util.io.Getline.getlineCall(Getline.java:39)
	at org.jruby.RubyIO.each(RubyIO.java:3533)
	at java.lang.invoke.LambdaForm$DMH/0x000000080129d400.invokeVirtual(java.base@17.0.6/LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/0x00000008012ac400.invoke(java.base@17.0.6/LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/0x00000008012f1400.tryFinally(java.base@17.0.6/LambdaForm$MH)
	at java.lang.invoke.DelegatingMethodHandle$Holder.delegate(java.base@17.0.6/DelegatingMethodHandle$Holder)
	at java.lang.invoke.LambdaForm$MH/0x000000080128ac00.guard(java.base@17.0.6/LambdaForm$MH)
	at java.lang.invoke.DelegatingMethodHandle$Holder.delegate(java.base@17.0.6/DelegatingMethodHandle$Holder)
	at java.lang.invoke.LambdaForm$MH/0x000000080128ac00.guard(java.base@17.0.6/LambdaForm$MH)
	at java.lang.invoke.Invokers$Holder.linkToCallSite(java.base@17.0.6/Invokers$Holder)
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_6_dot_0.gems.logstash_minus_input_minus_syslog_minus_3_dot_6_dot_0.lib.logstash.inputs.syslog.RUBY$method$tcp_receiver$0(/usr/share/logstash/vendor/bundle/jruby/2.6.0/gems/logstash-input-syslog-3.6.0/lib/logstash/inputs/syslog.rb:235)
```

It seems there is contention in fetching a selector from the selector pool, that is not observed in previous versions of JRuby/Logstash. Looking at the history, while it doesn't seem likely, could it be related to https://github.com/jruby/jruby/pull/6569 ?

**Comment by headius** [comment]
Ok I don't see anything very recent that would have changed behavior here, but I have a small patch that might reduce this contention a bit. SelectorPool could probably do better with some non-blocking collection love.

I'll push a PR. @jsvd are you able to reproduce to test this?

**Comment by jsvd** [comment]
thanks for the super fast response.

>  @jsvd are you able to reproduce to test this?

unfortunately no, I've been banging my head on this for a couple days now, both locally on my laptop and with a few vms, but can't replicate it. the report and stacktrace came from 3 individual users seeing the same thing.

**Comment by headius** [comment]
Sorry, I missed that you cannot reproduce. If you are at least able to provide a complete thread dump it would confirm that my fix would help (we should see one thread sitting in the retrieveFromPool method. My guess is that creating the new selectors is slowing down the process of putting them back, since both happen under lock. The new selector creation does not have to happen under lock.

See #7806.

**Comment by headius** [comment]
@jsvd Lots of stuff has changed from 9.2.20.1 😄

Most of the code in SelectorPool has not changed in a long time. However code that consumes the selectors may have changed in 9.3 to do a better job of putting back selectors after use. If that's the culprit, this should help.

**Comment by jsvd** [comment]
The only way I've been able to replicate this is to have multiple threads using Socket.each + "large" messages:

server.rb

```ruby
require 'logger'
require 'socket'

@logger = Logger.new(STDOUT)
@logger.level = Logger::INFO
@count = java.util.concurrent.atomic.AtomicInteger.new(0)

thread = Thread.new do
  @host = "127.0.0.1"
  @port = 3333
  @logger.info("Starting syslog tcp listener. address => #{@host}:#{@port}")
  @tcp = TCPServer.new(@host, @port)
  @tcp.do_not_reverse_lookup = true

  loop do
    socket = @tcp.accept

    Thread.new(socket) do |socket|
      peer_addr = socket.peeraddr
      ip, port = peer_addr[3], peer_addr[1]

      #@logger.info("new connection :client => #{ip}:#{port}")
      Thread.current.name = "input|syslog|tcp|#{ip}:#{port}}"

      socket.each {|line| @count.incrementAndGet }
    end
  end
end
loop { @logger.info("count: #{@count.longValue}"); sleep 10 }
```

client.rb:

```ruby
# ulimit -n 4000
# JRUBY_OPTS="-J-Xms4g -J-Xmx4g" rvm jruby-9.2.21.0 do ruby client.rb 127.0.0.1 3333 200
# JRUBY_OPTS="-J-Xms4g -J-Xmx4g" rvm jruby-9.3.10.0 do ruby client.rb 127.0.0.1 3333 200
require 'logger'
require 'socket'

@logger = Logger.new(STDOUT)
@logger.level = Logger::INFO
msg = "DEBUG OF: Generic HTTP Transfer: 6.53%: Download starts in 171 secondsInternal BAR-NG-VF500 cofsd: [0000000] "*20
data = "Tue Mar 04 15:57:06 2020: <15>Mar  4 15:53:03 BAR-NG-VF500 BAR-NG-VF500#{msg}\n"

threads = (ARGV[2].to_i).times.each_with_index.map do |i|
  Thread.new do
    sleep rand # stagger first write
    loop do
      @logger.info("opening client #{i}")
      socket = TCPSocket.new ARGV[0], ARGV[1].to_i

      loop do
        begin
          payload = "#{data*20}"
          while payload && payload.bytesize > 0
            written_bytes_size = socket.syswrite(payload)
            payload = payload.byteslice(written_bytes_size..-1)
          end
          sleep 0.01 # ~100 e/s
        rescue => e
          socket.close rescue nil
          break
        end

      end
    end

    socket.close rescue nil
  end
end
threads.each(&:join)
```

With both 9.2.21.0 and 9.3.10.0 I can see blocked threads on the same place as the reported users:

<details>
  <summary>see stacktrace</summary>

```
"Ruby-0-Thread-3@input|syslog|tcp|127.0.0.1:65295}: :1" #18 daemon prio=5 os_prio=31 cpu=488.12ms elapsed=12.50s tid=0x000000012f812400 nid=0x7207 waiting for monitor entry  [0x000000016de51000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.jruby.util.io.SelectorPool.returnToPool(org.jruby.dist/SelectorPool.java:137)
	- waiting to lock <0x0000000700001790> (a org.jruby.util.io.SelectorPool)
	at org.jruby.util.io.SelectorPool.put(org.jruby.dist/SelectorPool.java:100)
	at org.jruby.RubyThread.select(org.jruby.dist/RubyThread.java:2163)
	- locked <0x000000076795cac0> (a java.lang.Object)
	at org.jruby.RubyThread.select(org.jruby.dist/RubyThread.java:2053)
	at org.jruby.util.io.OpenFile.readInternal(org.jruby.dist/OpenFile.java:1394)
	at org.jruby.util.io.OpenFile.fillbuf(org.jruby.dist/OpenFile.java:1295)
	at org.jruby.util.io.OpenFile.getlineFast(org.jruby.dist/OpenFile.java:1635)
	at org.jruby.RubyIO.getlineImpl(org.jruby.dist/RubyIO.java:716)
	at org.jruby.RubyIO.access$100(org.jruby.dist/RubyIO.java:122)
	at org.jruby.RubyIO$4.getline(org.jruby.dist/RubyIO.java:2492)
	at org.jruby.RubyIO$4.getline(org.jruby.dist/RubyIO.java:2487)
	at org.jruby.util.io.Getline.getlineCall(org.jruby.dist/Getline.java:129)
	at org.jruby.util.io.Getline.getlineCall(org.jruby.dist/Getline.java:39)
	at org.jruby.RubyIO.each(org.jruby.dist/RubyIO.java:3533)
	at org.jruby.RubyIO$INVOKER$i$each.call(org.jruby.dist/RubyIO$INVOKER$i$each.gen)
	at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(org.jruby.dist/CachingCallSite.java:362)
	at org.jruby.runtime.callsite.CachingCallSite.call(org.jruby.dist/CachingCallSite.java:154)
	at org.jruby.runtime.callsite.CachingCallSite.callIter(org.jruby.dist/CachingCallSite.java:161)
	at Users.joaoduarte.investigations.tcp_performance.server.invokeOther7:each(/Users/joaoduarte/investigations/tcp_performance/server.rb:25)
	at Users.joaoduarte.investigations.tcp_performance.server.RUBY$block$\=\^main\_$2(/Users/joaoduarte/investigations/tcp_performance/server.rb:25)
	at java.lang.invoke.DirectMethodHandle$Holder.invokeStatic(java.base@17.0.4.1/DirectMethodHandle$Holder)
	at java.lang.invoke.LambdaForm$MH/0x000000080062c400.invoke(java.base@17.0.4.1/LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/0x000000080062d400.invokeExact_MT(java.base@17.0.4.1/LambdaForm$MH)
	at org.jruby.runtime.CompiledIRBlockBody.callDirect(org.jruby.dist/CompiledIRBlockBody.java:141)
	at org.jruby.runtime.IRBlockBody.call(org.jruby.dist/IRBlockBody.java:64)
	at org.jruby.runtime.IRBlockBody.call(org.jruby.dist/IRBlockBody.java:58)
	at org.jruby.runtime.Block.call(org.jruby.dist/Block.java:143)
	at org.jruby.RubyProc.call(org.jruby.dist/RubyProc.java:309)
	at org.jruby.internal.runtime.RubyRunnable.run(org.jruby.dist/RubyRunnable.java:107)
	at java.lang.Thread.run(java.base@17.0.4.1/Thread.java:833)
```

</details>

With your PR I still see the blocked threads but in another location:

<details>
  <summary>see stacktrace</summary>

```
"Ruby-0-Thread-198@input|syslog|tcp|127.0.0.1:65073}: :1" #213 daemon prio=5 os_prio=31 cpu=7098.57ms elapsed=165.43s tid=0x000000012d0d3a00 nid=0x2a403 waiting for monitor entry  [0x00000002b2ad1000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.jruby.RubyBasicObject.dataGetStruct(org.jruby.dist/RubyBasicObject.java:1090)
	- waiting to lock <0x0000000700531410> (a org.jruby.RubyArgsFile)
	at org.jruby.RubyArgsFile$ArgsFileData.getArgsFileData(org.jruby.dist/RubyArgsFile.java:199)
	at org.jruby.util.io.OpenFile.incrementLineno(org.jruby.dist/OpenFile.java:1651)
	at org.jruby.util.io.OpenFile.getlineFast(org.jruby.dist/OpenFile.java:1639)
	at org.jruby.RubyIO.getlineImpl(org.jruby.dist/RubyIO.java:716)
	at org.jruby.RubyIO.access$100(org.jruby.dist/RubyIO.java:122)
	at org.jruby.RubyIO$4.getline(org.jruby.dist/RubyIO.java:2492)
	at org.jruby.RubyIO$4.getline(org.jruby.dist/RubyIO.java:2487)
	at org.jruby.util.io.Getline.getlineCall(org.jruby.dist/Getline.java:129)
	at org.jruby.util.io.Getline.getlineCall(org.jruby.dist/Getline.java:39)
	at org.jruby.RubyIO.each(org.jruby.dist/RubyIO.java:3533)
	at org.jruby.RubyIO$INVOKER$i$each.call(org.jruby.dist/RubyIO$INVOKER$i$each.gen)
	at org.jruby.runtime.callsite.CachingCallSite.call(org.jruby.dist/CachingCallSite.java:152)
	at org.jruby.runtime.callsite.CachingCallSite.callIter(org.jruby.dist/CachingCallSite.java:161)
	at Users.joaoduarte.investigations.tcp_performance.server.invokeOther7:each(/Users/joaoduarte/investigations/tcp_performance/server.rb:25)
	at Users.joaoduarte.investigations.tcp_performance.server.RUBY$block$\=\^main\_$2(/Users/joaoduarte/investigations/tcp_performance/server.rb:25)
	at java.lang.invoke.DirectMethodHandle$Holder.invokeStatic(java.base@17.0.4.1/DirectMethodHandle$Holder)
	at java.lang.invoke.LambdaForm$MH/0x0000000800661800.invoke(java.base@17.0.4.1/LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/0x000000080065a400.invokeExact_MT(java.base@17.0.4.1/LambdaForm$MH)
	at org.jruby.runtime.CompiledIRBlockBody.callDirect(org.jruby.dist/CompiledIRBlockBody.java:141)
	at org.jruby.runtime.IRBlockBody.call(org.jruby.dist/IRBlockBody.java:64)
	at org.jruby.runtime.IRBlockBody.call(org.jruby.dist/IRBlockBody.java:58)
	at org.jruby.runtime.Block.call(org.jruby.dist/Block.java:143)
	at org.jruby.RubyProc.call(org.jruby.dist/RubyProc.java:309)
	at org.jruby.internal.runtime.RubyRunnable.run(org.jruby.dist/RubyRunnable.java:107)
	at java.lang.Thread.run(java.base@17.0.4.1/Thread.java:833)
```

</details>

I'm not sure if what I'm replicating is the customer issue, since I still see blocked threads on JRuby 9.2, but just wanted to give you this update.

**Comment by headius** [comment]
Your new trace does not seem problematic to me; Socket#each uses IO#each which does set up the ArgsFile stuff and that requires synchronization as it is shared across threads. The patches I provided eliminate SelectorPool from contention altogether.

I do share your concern about this case also showing up on 9.2.21. When your users reported high levels of contention, was that contention slowing things down or did they just happen to notice lots of threads waiting on locks or blocking calls? Heavy IO will always have some threads blocking on resource acquisition or on the IO calls themselves.

I will spend a few minutes looking into the ArgsFile stuff, though, since it probably has not been reworked for better concurrency in many years.

**Comment by headius** [comment]
The use of `synchronized` on the `dataGetStruct` method does not appear to be necessary. It dates all the way back to the introduction of the method in this commit in 2006: https://github.com/jruby/jruby/commit/308debcc1308e8e986d0226f0c4ac3aab13402a6

I see no reason to have it in place anymore, as it uses the VariableTableManager which has been vetted for concurrent use.

**Comment by headius** [comment]
@enebo These fixes should port forward to 9.4, either for 9.4.3 or with some bake time into 9.4.4.

**Comment by headius** [comment]
I've merged the patches from #7806 and with a short exercising code similar to your most recent stack traces, I see full utilization of cores and no obvious points of contention. An additional minor change in #7806 reduces allocation during `puts` calls which invokes `write` with two arguments.

These changes have been cherry-picked for JRuby 9.4 in #7823.
